### PR TITLE
Secret - from hash function, also validate data

### PR DESCRIPTION
**Pull request by NikVolf**

### CODE REVIEW

**Comment by tomusdrw** [comment]
I think we should validate the hash before turning it into secret: https://github.com/apoelstra/rust-secp256k1/blob/master/src/key.rs#L86
That implementation was recently changed to avoid blindly converting hash to slice and rather pass it through `Secret::from_slice` or `Secret::from_hash`

**Comment by NikVolf** [comment]
Hmm 
Why `Secret::from_slice` does not validate anything then?

I thought of this as a shortcut of `from_slice` actually

**Comment by tomusdrw** [comment]
Darn, you're absolutely right. I wanted to use `secp256_k1::key::SecretKey::from_slice` there, but probably missed that when working on it:
https://github.com/ethcore/parity/blob/master/ethkey/src/secret.rs#L43
It should definitely be verified, good that you've noticed it.

**Comment by NikVolf** [comment]
@tomusdrw 
ok i'll add to both here

File: ethkey/src/secret.rs [hunk 1]
```diff
@@ -54,17 +63,11 @@ impl FromStr for Secret {
 
 impl From<key::SecretKey> for Secret {
 	fn from(key: key::SecretKey) -> Self {
-		Self::from_slice(&key[0..32])
+		Self::from_slice_unchecked(&key[0..32])
 			.expect("`key::SecretKey` is valid (no way to construct invalid one); qed")
```
Comment (tomusdrw) [inline ethkey/src/secret.rs]: Could get rid of this expect by moving `.len() != 32` check to `from_slice`

Comment (NikVolf) [inline ethkey/src/secret.rs]: replaced it with just assert
since `SecretKey` have got this check already

File: ethkey/src/secret.rs [hunk 2]
```diff
@@ -33,14 +33,21 @@ impl fmt::Debug for Secret {
...
+		Secret { inner: h }
+	}
+
+	pub fn from_slice(key: &[u8]) -> Result<Self, Error> {
+		let secret = key::SecretKey::from_slice(&super::SECP256K1, key)?;
+		Ok(secret.into())
+	}
+
+	pub fn from_hash(key: &H256) -> Result<Self, Error> {
+		Self::from_slice(&**key)
```
Comment (rphmeier) [inline ethkey/src/secret.rs]: Is this function necessary? From an abstract standpoint getting a secret from a hash doesn't make a whole lot of sense and from a code standpoint the `H256` already has a `Deref` impl.

Comment (NikVolf) [inline ethkey/src/secret.rs]: though it was from what pr started, now indeed there is no need in it

was hoping to use `into()` in few cases, now not :)

File: ethkey/src/secret.rs [hunk 3]
```diff
@@ -33,14 +33,17 @@ impl fmt::Debug for Secret {
...
 
 		let mut h = H256::default();
 		h.copy_from_slice(&key[0..32]);
-		Ok(Secret { inner: h })
+		Secret { inner: h }
+	}
+
+	pub fn from_slice(key: &[u8]) -> Result<Self, Error> {
+		let secret = key::SecretKey::from_slice(&super::SECP256K1, key)?;
+		Ok(secret.into())
```
Comment (rphmeier) [inline ethkey/src/secret.rs]: could also `.map(Into::into)`

Comment (tomusdrw) [inline ethkey/src/secret.rs]: with `.map_err(Into::into)` :/

Comment (rphmeier) [inline ethkey/src/secret.rs]: ah, yeah. probably cleaner this way then :)

**Pull request merged**